csharp_dotnetcore/3.Welcome-Users/WelcomeUserBot.cs

```diff
@@ -0,0 +1,146 @@
+    /// For each Activity received, a new instance of this class is created.
+    /// Objects that are expensive to construct, or have a lifetime.
+    /// beyond the single Turn, should be carefully managed and chached - see Startup.cs.
+    /// </summary>
+    public class WelcomeUserBot : IBot
+    {
+        // generic message sent to user
+        private const string _genericMessage = "This is a simple Welcome Bot sample.You can say \'intro\' to see the introduction card. If you are running this bot in the Bot Framework emulator, press the \'Start Over\' button to simulate user joining a bot or a channel";
+
+        // the bot state accessor object. Use this to access specific state properties
+        private readonly WelcomeUserStateAccessors _welcomeUserStateAccessors;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="WelcomeUserBot"/> class.
+        /// Initalize WelcomeUserBot.
+        /// </summary>
+        /// <param name="statePropertyAccessor"> Bot state accessor object.</param>
+        public WelcomeUserBot(WelcomeUserStateAccessors statePropertyAccessor)
+        {
+            _welcomeUserStateAccessors = statePropertyAccessor ?? throw new System.ArgumentNullException("state accessor can't be null");
+        }
+
+        /// <summary>
+        /// Every Conversation turn for our WelcomeUser Bot will call this method, including
+        /// any type of activiaties such as ConversationUpdate or ContactRelationUpdate which
+        /// are sent when a user joins a conversation.
+        /// This bot doesn't use any dialogs; it's "single turn" processing, meaning a single
+        /// request and response, with no stateful conversation between user input.
+        /// This bot does use UserState to keep track of first message users send to the bot.
+        /// </summary>
+        /// <param name="turnContext">A <see cref="ITurnContext"/> containing all the data needed
+        /// for processing this conversation turn. </param>
+        /// <param name="cancellationToken">(Optional) A <see cref="CancellationToken"/> that can be used by other objects
+        /// or threads to receive notice of cancellation.</param>
+        /// <returns>A task that represents the work queued to execute.</returns>
+        public async Task OnTurnAsync(ITurnContext turnContext, CancellationToken cancellationToken = new CancellationToken())
+        {
+            // use state accessor to extract the didBotWelcomeUser flag
+            var didBotWelcomeUser = await _welcomeUserStateAccessors.DidBotWelcomedUser.GetAsync(turnContext, () => false);
+
+            if (turnContext.Activity.Type == ActivityTypes.Message && turnContext.Responded == false)
+            {
+                // Your bot should proactively send a welcome message to a personal chat the first time
+                // (and only the first time) a user initiates a personal chat with your bot.
```

Comment:
Is this guidance correct? I don't think it is. The guidance is (I believe) that any time a user joins a new conversation, we should send welcome. For example, given this guidance, a greeting message would only ever be sent once, even if the last time I chatted with the bot was 6 months ago.